argon2_out_of_process_support_sync.go

```diff
@@ -0,0 +1,247 @@
+// the process exits too).
+//
+// The specified timeout determines how long this function will wait before
+// aborting its attempt to acquire the lock. If set to 0, the function will
+// only perform a single attempt.
+func acquireArgon2OutOfProcessHandlerSystemLock(timeout time.Duration) (release func(), err error) {
+	var lockFile *os.File // The opened lock file
+
+	// Ensure that we close the open lockFile descriptor (if there is one)
+	// on error paths. Note that this does leave a lock file laying around,
+	// but this isn't a problem and unlinking it can only happen inside of
+	// an exclusive lock without breaking the locking contract anyway.
+	defer func() {
+		if err == nil || lockFile == nil {
+			return
+		}
+		if err := lockFile.Close(); err != nil {
+			fmt.Fprintf(argon2SysLockStderr, "Cannot close argon2 lock file descriptor on error: %v", err)
+		}
+	}()
+
+	timeoutTimer := time.NewTimer(timeout) // Begin the request timeout timer
+	triedOnce := false                     // Handle the case of timeout == 0
+
+	// Run a loop to try to acquire the lock.
+	for {
+		if triedOnce {
+			// If the loop has executed at least once, make sure that
+			// the timeout hasn't expired.
+			select {
+			case <-timeoutTimer.C:
+				// The timeout has expired.
+				return nil, errArgon2OutOfProcessHandlerSystemLockTimeout
+			default:
+				// continue trying
+			}
+		}
+		triedOnce = true
+
+		// Make sure that we close the lock file left open from the previous
+		// attempt, if there is one. Note that this does leave a lock file
+		// laying around, but this isn't a problem and unlinking it can only
+		// happen inside of an exclusive lock without breaking the locking
+		// contract anyway.
+		if lockFile != nil {
+			if err := lockFile.Close(); err != nil {
+				return nil, fmt.Errorf("cannot close lock file from previous attempt before starting new attempt: %w", err)
+			}
+		}
+
+		// Attempt to open the lock file for writing.
+		lockFile, err = os.OpenFile(paths.Argon2OutOfProcessHandlerSystemLockPath, os.O_RDWR|os.O_CREATE, 0600)
+		if err != nil {
+			// No error is expected here.
+			return nil, fmt.Errorf("cannot open lock file for writing: %w", err)
+		}
+
+		// Grab information about the lock file we just opened, via its descriptor.
+		var lockFileSt unix.Stat_t
+		if err := unix.Fstat(int(lockFile.Fd()), &lockFileSt); err != nil {
+			// No error is expected here
+			return nil, fmt.Errorf("cannot obtain lock file info from open descriptor: %w", err)
+		}
+
+		// Attempt to acquire an exclusive, non-blocking, advisory lock.
+		if err := unix.Flock(int(lockFile.Fd()), unix.LOCK_EX|unix.LOCK_NB); err != nil {
+			// We failed to acquire the lock.
+			if os.IsTimeout(err) {
+				// The EWOULDBLOCK case. Someone else already has a lock on the
+				// file we have opened. Try again with a 10ms backoff time,
+				// unless the request timeout fires first.
+				select {
+				case <-time.NewTimer(10 * time.Millisecond).C:
```

Comment:
isn't this a bit too aggressive? how long does it take usually to compute the actual KDF?

Reply:
At this point, we have no idea how long it will be before we can acquire the lock, so it's a bit of a trade-off between being aggressive and waiting longer than is necessary, where the lock ends up unheld for longer for a period. I'm not sure what time to use for the backoff here tbh - the other option is to make the call to `flock` blocking, but then there's no way for it to be able to timeout.